Acquire CurrentResourceOwner in resgroup/queue code

amilkh opened the pull request:
A few methods used in the resgroup/queue code require having
CurrentResourceOwner set when they are called. This commit introduces
two new functions DefaultResourceOwnerAcquire() and
DefaultResourceOwnerRelease() which can be used to do this.

Placed in resource_manager.c/h instead of the more appropriate upstream
resowner.c/h, as:
1. Modifying resowner.c/h would cause conflicts when cherry-picking
   commits from newer versions of Postgres.
2. resource_manager.c/h are small components shared between resource
   group and resource queue code.

Signed-off-by: David Sharp <dsharp@pivotal.io>
Signed-off-by: Ben Christel <bchristel@pivotal.io>

amilkh commented:
More context:

For the `DefaultResourceOwnerAcquire()` and `DefaultResourceOwnerRelease()` methods, we searched the gpdb codebase, and seems these are the only places they should be used:
1. resgroupcmds.c (resource group code): `GetResGroupIdForRole()` and `GetResGroupIdForName()`
2. resscheduler.c (resource queue code): `InitResQueues()` and `setResQueueId()`

I considered the below options and went with #4:
1. resowner.c/h: this would be ideal, but it is Postgres code that we should not modify to avoid upstream conflicts
https://github.com/greenplum-db/gpdb/blob/master/src/include/utils/resowner.h
https://github.com/postgres/postgres/blob/master/src/include/utils/resowner.h
2. resgroupcmds.c/h: strange because the methods are used by resscheduler
3. resscheduler.c/h: strange because the methods are used by resgroupcmds
4. resource_manager.c/h: a bit strange because this is for the `gp_resource_manager` GUC (https://gpdb.docs.pivotal.io/500/ref_guide/config_params/guc-list.html#gp_resource_manager), but probably best because it is shared code by resource groups and queues
5. new file (i.e. resource_manager_helpers.c/h): probably not worth adding these new files just for these two methods

hlinnaka commented:
This looks fishy. CurrentResourceOwner is always set, when you're in a transaction. These functions are accessing catalogs, which is also not OK when you're not in a transaction. In other words, none of this should be needed. 

Looking at the callers of these, AssignResGroupOnMaster() is called very early in StartTransaction(). Why is it called so early?

dsharp-pivotal commented:
>This looks fishy. CurrentResourceOwner is always set, when you're in a transaction. These functions are accessing catalogs, which is also not OK when you're not in a transaction. In other words, none of this should be needed.
>
>Looking at the callers of these, AssignResGroupOnMaster() is called very early in StartTransaction(). Why is it called so early?

You're right all of this is before the transaction is created. Can you explain more why it is "not OK"? It certainly seems to work.

As for why it was done this way, @xiong-gang can explain more, but... Transactions are assigned to resource groups in part because assigning outside a transaction avoids a possible deadlock between ResGroupLock and table locks.

amilkh commented:
@hlinnaka 
Also, to add onto what David said, the resource group that a transaction will use is determined before the transaction starts running. On the open source side, this is (already, before this PR existed) done with `decideResGroupId()` -> `GetResGroupIdForRole()` which looks at a cache of `pg_authid` to pull the resource group that corresponds to the role running that transaction.

hlinnaka commented:
On 11/20/2017 08:38 PM, David Sharp wrote:
>> This looks fishy. CurrentResourceOwner is always set, when you're
>> in a transaction. These functions are accessing catalogs, which is
>> also not OK when you're not in a transaction. In other words, none
>> of this should be needed.
>> 
>> Looking at the callers of these, AssignResGroupOnMaster() is called
>> very early in StartTransaction(). Why is it called so early?
> 
> You're right all of this is before the transaction is created. Can
> you explain more why it is "not OK"? It certainly seems to work.

Well, you don't have the resource owner machinery set up yet, for 
starters, as you found out :-). You also don't have a per-transaction 
memory context yet, so any leaked memory will not be cleaned up at 
end-of-transaction, like it normally would.

Relcache invalidation is in between-xacts state, so I'm not sure what 
would happen if a relcache invalidation event for the catalog relation 
arrives while you're accessing it. Might work, but I wouldn't bet on it.

You cannot acquire a snapshot without a transaction, or you'll confuse 
the snapshot manager. The catalog access in this case is done with 
SnapshotNow, so that doesn't require a snapshot, but I'd still be 
uncomfortable by running with an out-of-date RecentGlobalXmin. Maybe 
it's OK, because it's generally safe to have an older RecentGlobalXmin 
than necessary, but I would worry about behavior around wraparound, if a 
backend process lingers unused for a very long time. The backend 
startup, in InitPostgres, acquires a snapshot for that reason. (Note 
that InitPostgres opens a transaction while it performs catalog 
lookups.) Also, SnapshotNow was removed in PostgreSQL 9.4, so we'd 
surely need to revisit this at that point anyway.

Those fears might turn out to be non-issues, or you might be able to 
work around them, like you worked around the lack of 
CurrentResourceOwner. But it's going to be an uphill battle, and it's 
likely to get broken as we merge with later versions of PostgreSQL, 
because PostgreSQL assumes that you always have a transaction open when 
doing catalog lookups.

- Heikki

hlinnaka commented:
On 11/20/2017 08:56 PM, Amil Khanzada wrote:
> @hlinnaka Also, to add onto what David said, the resource group that
> a transaction will use is determined before the transaction starts
> running.

Yes, but why? I understand that you don't want to run queries before 
you've decided the resource group to use, but there shouldn't be any 
harm in opening the transaction first. An open transaction doesn't use 
much resources.

I'm basically suggesting that we move the call to 
AssignResGroupOnMaster() from the beginning of StartTransaction() to the 
end of StartTransaction(). If you look at the things that are done in 
StartTransaction(), none of of those things should consume any 
meaningful amount of resources.

- Heikki

amilkh commented:
@hlinnaka wow, great points! I didn't consider all of that.

Looks like this PR is blocked pending on the discussion between you and MPP team whether the below change should happen:
> I'm basically suggesting that we move the call to
AssignResGroupOnMaster() from the beginning of StartTransaction() to the
end of StartTransaction(). If you look at the things that are done in
StartTransaction(), none of of those things should consume any
meaningful amount of resources.

If that change does happen, this PR becomes irrelevant because our code path will now be in a transaction, so `CurrentResourceOwner` will already be set (i.e. we don't need to acquire a new one).

@xiong-gang @pivotal-ning-yu what do you guys think?

xiong-gang commented:
@hlinnaka Thanks for the detailed explanation.
The reason we put `AssignResGroupOnMaster()` at the beginning of `StartTransaction()` is `AssignResGroupOnMaster()` may wait on resource group slot, if we already acquired some locks before that, there might be a deadlock. 
So what locks might be acquired in `StartTransaction()`, xid lock seems fine, because if it's waiting for resource group slot at the end of `StartTransaction`, other transaction won't wait for its xid lock.

amilkh commented:
@xiong-gang thanks for explaining!
What I understand from what you are saying is that resource groups have concurrency/memory/cpu limits, so it might be the case that a transaction should not be allowed to `StartTransaction()` and acquire locks, thus `AssignResGroupOnMaster()` is called before to validate the resource group which should be used (based on role in `pg_authid` in the open source code).
I guess this can be when the `waiting_reason` column of `pg_stat_activity` says `resgroup`.
https://gpdb.docs.pivotal.io/520/ref_guide/system_catalogs/pg_stat_activity.html

amilkh commented:
@hlinnaka what do you think? Are the risks you mentioned (stale cache, conflict with future versions of Postgres, etc.) that critical that we should block this PR and also rework the order of `StartTransaction()` and `AssignResGroupOnMaster()`?

hlinnaka commented:
On 11/21/2017 07:58 PM, Amil Khanzada wrote:
> @hlinnaka what do you think? Are the risks you mentioned (stale
> cache, conflict with future versions of Postgres, etc.) that critical
> that we should block this PR and also rework the order of
> `StartTransaction()` and `AssignResGroupOnMaster()`?

Yes.

- Heikki

amilkh commented:
I see, thanks for bringing these important points up @hlinnaka!
David and I have started a thread on the mailing list.
"Ordering of AssignResGroupOnMaster() and StartTransaction()"

Please take a look as well, @xiong-gang @pivotal-ning-yu .

amilkh commented:
Based on discussion in the email thread on gpdb-dev, the MPP team will make a change so that the a transaction is already started when `AssignResGroupOnMaster()` is called. This will mean that methods that are involved in the resource group decision making and access catalogs like `GetResGroupIdForRole()` will have the `CurrentResourceOwner` set. Once that is done, we can remove some of the existing code that did `ResourceOwnerCreate` like below:
https://github.com/greenplum-db/gpdb/pull/3924/files#diff-1c336ba9e49f7d55a9e86427c96e1637L533

@xiong-gang @hlinnaka For now, is it alright if we merge this PR as is? We need it for some of our code which calls `GetResGroupIdForName()`. Once the MPP team has made their change (not sure of when) and assuming it resolves this problem, we can remove the calls to `DefaultResourceOwnerAcquire()` and `DefaultResourceOwnerRelease()`.

hlinnaka commented:
On 11/22/2017 08:44 PM, Amil Khanzada wrote:
> Based on discussion in the email thread on gpdb-dev, the MPP team
> will make a change so that the a transaction is already started when
> `AssignResGroupOnMaster()` is called. This will mean that methods
> that are involved in the resource group decision making and access
> catalogs like `GetResGroupIdForRole()` will have the
> `CurrentResourceOwner` set. Once that is done, we can remove some of
> the existing code that did `ResourceOwnerCreate` like below: 
> https://github.com/greenplum-db/gpdb/pull/3924/files#diff-1c336ba9e49f7d55a9e86427c96e1637L533
>
>  @xiong-gang @hlinnaka For now, is it alright if we merge this PR as
> is? We need it for some of our code which calls
> `GetResGroupIdForName()`. Once the MPP team has made their change
> (not sure of when) and assuming it resolves this problem, we can
> remove the calls to `DefaultResourceOwnerAcquire()` and
> `DefaultResourceOwnerRelease()`.

Why don't you just move the AssignResGroupOnMaster() call to later in 
StartTransaction() now, and skip the back and forth? If that what you're 
going to do in the end anyway, might as well do it straight away.

- Heikki

amilkh commented:
On Wed, Nov 22, 2017 at 11:35 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> Why don't you just move the AssignResGroupOnMaster() call to later in
> StartTransaction() now, and skip the back and forth? If that what you're
> going to do in the end anyway, might as well do it straight away.
>
>
> - Heikki
>

Good idea! I'm just not sure of the effort involved in that.
@xiong-gang if it's something easy, could you please tell us what the code
change would be?

PS: We're off for Thursday/Friday (Thanksgiving), but will be back on
Monday.

xiong-gang commented:
@amilkh I think you could just move this call to the end of StartTransaction(), it should be fine, we can fix it later when we run into any problem.

amilkh commented:
Thanks Gang!

I'll try it out next week when I'm back in office unless you want to do it
sooner.

Let's see if Concourse still passes with that change!

2017年11月23日木曜日、xiong-gang<notifications@github.com>さんは書きました:

> @amilkh <https://github.com/amilkh> I think you could just move this call
> to the end of StartTransaction(), it should be fine, we can fix it later
> when we run into any problem.
>

dsharp-pivotal commented:
@xiong-gang, we were looking at this comment above the call to `AssignResGroupOnMaster()` in `StartTransaction()`:

```
	/*		
	* Acquire a resource group slot.		
	 *		
	 * AssignResGroupOnMaster() might throw error, so call it before touch		
	 * transaction state.		
	 * Slot is successfully acquired when AssignResGroupOnMaster() is returned,		
	 * this slot will be release when transaction is committed or abortted,		
	 * so don't error out before transaction state is set to TRANS_START.		
	 */
```

Are we still concerned about throwing an error if we call `AssignResGroupOnMaster` at the end of `StartTransaction`, when the transaction state has presumably been modified? @amilkh and @benchristel updated the comment a bit, but can you look to see if it is still accurate?

dsharp-pivotal commented:
We have a dev pipeline up with this branch. It's passing everything so far.

https://gpdb-dev.data.pivotal.ci/teams/main/pipelines/dev:amilkh_resgroup_default_resource_owner

xiong-gang commented on src/backend/access/transam/xact.c:
```diff
@@ -2437,6 +2425,18 @@ StartTransaction(void)
+	 * call it after touching the transaction state.
+	 * Slot is successfully acquired when AssignResGroupOnMaster() is returned,
+	 * this slot will be release when transaction is committed or aborted,
+	 * so don't error out before transaction state is set to TRANS_START.
```
@benchristel 
I think we can remove this comment now.
As the transaction state is already set to TRANS_INPROGRESS, anything goes wrong will trigger `AbortTransaction`, and the slot will be released there.

benchristel replied:
👍  Thanks, updated it!

benchristel commented:
The `compile_*`, `icw_*`, and `cs_*` jobs have all passed on [our dev pipeline](https://gpdb-dev.data.pivotal.ci/teams/main/pipelines/dev:amilkh_resgroup_default_resource_owner). @xiong-gang can we merge this?

xiong-gang commented:
@benchristel yes, go ahead, please.

amilkh commented:
@xiong-gang sorry to keep dragging this on, but a few updates:
1. (code change) Removed the code in `GetResGroupIdForRole()` which acquires the `CurrentResourceOwner` before querying `pg_authid`
2. Updated the code comment in `StartTransaction()`
3. Updated the commit message

Our [dev pipeline](https://gpdb-dev.data.pivotal.ci/teams/main/pipelines/dev:amilkh_resgroup_default_resource_owner) which runs most jobs on Concourse is green with these changes, so we'll merge this tomorrow if no objections.


Thanks all for your input, especially @hlinnaka for changing the flow of this PR entirely, from a new code addition to an enhancement on existing code!

xiong-gang commented:
@amilkh the change looks good, thanks.
